src/lib/math/pcurves/pcurves_impl.h

```diff
@@ -0,0 +1,1428 @@
+
+      static constexpr Self one() { return Self(Rep::one()); }
+
+      static constexpr Self from_word(W x) {
+         std::array<W, 1> v{x};
+         return Self::from_words(v);
+      }
+
+      template <size_t L>
+      static constexpr Self from_words(std::array<W, L> w) {
+         if constexpr(L == N) {
+            return Self(Rep::to_rep(w));
+         } else {
+            static_assert(L < N);
+            std::array<W, N> ew = {};
+            for(size_t i = 0; i != L; ++i) {
+               ew[i] = w[i];
+            }
+            return Self(Rep::to_rep(ew));
+         }
+      }
+
+      constexpr bool is_zero() const { return CT::all_zeros(m_val.data(), m_val.size()).as_bool(); }
+
+      constexpr bool is_nonzero() const { return !is_zero(); }
+
+      constexpr bool is_one() const { return (*this == Self::one()); }
+
+      constexpr bool is_even() const {
+         auto v = Rep::from_rep(m_val);
+         return (v[0] & 0x01) == 0;
+      }
+
+      friend constexpr Self operator+(const Self& a, const Self& b) {
+         std::array<W, N> t;
+         W carry = bigint_add<W, N>(t, a.value(), b.value());
+
+         std::array<W, N> r;
+         bigint_monty_maybe_sub<N>(r.data(), carry, t.data(), P.data());
+         return Self(r);
+      }
+
+      Self dbl() const {
+         std::array<W, N> t = value();
+         W carry = shift_left<1>(t);
+
+         std::array<W, N> r;
+         bigint_monty_maybe_sub<N>(r.data(), carry, t.data(), P.data());
+         return Self(r);
+      }
+
+      constexpr Self& operator+=(const Self& other) {
+         std::array<W, N> t;
+         W carry = bigint_add<W, N>(t, this->value(), other.value());
+         bigint_monty_maybe_sub<N>(m_val.data(), carry, t.data(), P.data());
+         return (*this);
+      }
+
+      friend constexpr Self operator-(const Self& a, const Self& b) { return a + b.negate(); }
+
+      friend constexpr Self operator*(uint8_t a, const Self& b) { return b * a; }
+
+      friend constexpr Self operator*(const Self& a, uint8_t b) {
+         // We assume b is a small constant and allow variable time
+         // computation
+
+         // In practice this function is called for 2, 3, 4, or 8
+
+         if(b == 2) {
+            return a.dbl();
+         } else if(b == 3) {
+            return a.dbl() + a;
+         } else if(b == 4) {
+            return a.dbl().dbl();
+         } else if(b == 8) {
+            return a.dbl().dbl().dbl();
+         } else {
+            Self z = Self::zero();
+            Self x = a;
+
+            while(b > 0) {
+               if(b & 1) {
+                  z = z + x;
+               }
+               x = x.dbl();
+               b >>= 1;
+            }
+
+            return z;
+         }
+      }
+
+      friend constexpr Self operator*(const Self& a, const Self& b) {
+         std::array<W, 2 * N> z;
+         comba_mul<N>(z.data(), a.data(), b.data());
+         return Self(Rep::redc(z));
```

Comment:
Nit: Perhaps implement this in terms of `operator*=` like:

```suggestion
         auto result = a;
         result *= b;
         return result;
```

... same for `operator+` and the others, of course. That would reduce code duplication a tiny bit.